[Testmerge?] TGUI View Variables panel

TheTrigg opened the pull request:
## About The Pull Request

Hey, maybe I shouldn't try to mess with TGUI with LITERALLY zero Javascript knowledge.
It compiles though. And eslint is happy.

_You will still probably want to gouge your eyes out, though._

For now, the original VV is intact.
Plan is to replace it once this one can do everything the old one can.

Still a lot to do. I mean, a lot.
- [x] Icon handling
- [x] File handling
- [x] Matrix handling
- [x] Proper list handling
- [x] Directions
- [x] (E) becomes click, (C) becomes shift click, (M) becomes shift-ctrl click
- [x] Basically the entirety of ui_act
- [x] Button to share a VV panel via admin chat, so all other admins can wonder how the clown broke the fabric of reality again
- [x] Add more useful dropdown options, like, a "show in team panel" to a /datum/team, or a simple "aheal" to a /mob/living.
- [ ] Make the JS side less garbage because holy shit I suck at this
- [ ] Make the byond side faster...?
- [ ] Find a way to get a proper dropdown panel that doesn't break itself OR the interface, and doesn't stay in the way when collapsed.
- [ ] Get Javascript to do nearest-neighbor pixel scaling somehow
- [ ] Get Labeledlist.Item to behave properly with bounding, or just completely ditch Labeledlist...?


I'll maybe do these on a follow-up PR. Not now though, else I'll go completely mad.
~~-  Matrix Manipulation (Modify Transform V2 I guess)~~
~~- [Entirety of a health manipulation screen: Shows all organs/implants, quick adding, healing/damaging, limb and species manipulation too. Erm.. Maybe on a separate project/pr?~~
~~-  "Watch" option - Constantly displays the value of desired vars in a category, like the MC monitoring category. Again, separate project? ...Is it even possible? Hell if I know... For now.~~
~~-  Ability to group vars by their inherited component~~
~~-  Come up with more ideas for this list. Open to suggestions, of course - No idea's too dumb! Most of the time.~~

## So uh, yeah, I need help
This list is big. It's an ambitious project.
I was not kidding when I said I had no JS knowledge. I never touched JS before outside of this, and another TGUI test project.
I am also pretty much a _big_ noob to coding in general, and know few DM coding standards, what's better, faster, more readable, etc.
This means that it'll take a lot of time for me to get this done alone.
Without help, it _will_ most likely get done bacause I'm a tenacious fuck with no life, but _it'll take a while._
Talk to me over discord, I guess: Trigg#0475

If you _do_ wanna help me, please, be pacient with my little slow brain ok thanks

## Why It's Good For The Game

![image](https://user-images.githubusercontent.com/36010999/88488068-2fdcd380-cf61-11ea-9424-7449d68dee95.png)
![image](https://user-images.githubusercontent.com/36010999/88488070-35d2b480-cf61-11ea-9948-da689132954a.png)
![image](https://user-images.githubusercontent.com/36010999/88488073-39663b80-cf61-11ea-8641-ec29e00d931d.png)
![image](https://user-images.githubusercontent.com/36010999/88488074-3cf9c280-cf61-11ea-83fc-3fde58ba0edc.png)



## Changelog
:cl: Trigg
add: TGUI VV panel, may your eyes burn no more!
add: Remember those (E) (C) (M) options? Well, (E) is now a normal click, (C) a shift click, and (M) a shift-ctrl click. Lookin' clean.
add: Two new dropdown options: Smite and Aheal.
add: Admins can now share VV panels via admin chat, so everyone can be on the same page when you say "The clown made a weapon that moves you to nullspace"
/:cl:

TheTrigg committed:
aight lemme get what ive done so far for the draft

TheTrigg committed:
dropdown placement is *kinda* done

TheTrigg committed:
ooh, that's a cool trick

TheTrigg committed:
scrollable sections because I need em

TheTrigg committed:
Dropdown menu now scrolls, neat

xboxnoob commented:
hey trigg can i have your discord tag i want to talk to you about this project

TheTrigg commented:
Oh. Right. I'm dumb. I asked for help but PROVIDED NO WAY FOR PEOPLE TO CONTACT ME.

Trigg#0475

TheTrigg committed:
y'know this is even better

TheTrigg committed:
directions; like 10% of ui_act

TheTrigg committed:
uhhh rotation's done

TheTrigg committed:
this whole system needs some extra OOPing

TheTrigg commented:
Uhhhh after some _"reinforcing"_ of the way you all review and accept PRs in https://github.com/BeeStation/BeeStation-Hornet/pull/2046, I highly doubt this one will be up to your code standards either. So I'll just not bother continuing this PR. Wish you all the best of luck going forward. See ya.

Crossedfall commented:
That applied specifically to hippie prs, as laid out by our guidelines. This pr is also substantially smaller. So, to be clear, we're happy to guide people through what needs to be done to get their prs approved. But, for massive projects or ports from notoriously different codebases, there is a higher expectation that has to be had for us to be able to work effectively. We sadly don't have the same resources a place like tg does.

If you're truly abandoning this out of spite, then while it'd be disappointing that is a decision you're free and welcome to make and hopefully we'll see you again at some point. If not, best of luck to you as well.

TheTrigg commented:
If you're telling me it was just specifically Hippie PRs and not this much worse dumpster fire of a "me" PR that's getting yeeted down, I mean alright I'll reopen this. There was no spite, just thought this would be shot down too.

TheTrigg committed:
this shit is all over the place
and I'll probably need to redo all this instead of attempting to adapt it

TheTrigg committed:
file and icon handling... sort of

TheTrigg committed:
aheal dropdown option
wow, much difficulty, big brain coder moment
also removed duplicated code I guess
oh, and ahealing should be logged to tickets and feedback tally now

TheTrigg committed:
the chad stylemistake vs the virgin trigg

TheTrigg committed:
thanks stylemistake

TheTrigg committed:
pretty sure ui_act is... done? kind of?
In a very, very, duct-taped way?

TheTrigg commented:
Surprisingly, this is almost at a stage where it can replace the old VV, even though I've only 100% completed two of the targets in the todo.

I'm having a temporary rimworld addiction so you won't be seeing many commits for a while.
I still fully intend to finish this.

TheTrigg committed:
"smite" dropdown option

TheTrigg committed:
forgot to stage this

TheTrigg committed:
damage buttons work and lists almost work properly

TheTrigg committed:
very responsible to commit without even testing

TheTrigg committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into trigg-sucks-at-javascript

TheTrigg committed:
pulled stuff

TheTrigg committed:
oh.

TheTrigg committed:
bug bad

TheTrigg committed:
smort

TheTrigg committed:
insert clever matrix quote here

TheTrigg commented:
Pretty sure feature parity with the old VV has been reached.
@xboxnoob will do the health manipulation stuff, though he's prioritizing his bounty; I'm still left with everything else.

TheTrigg committed:
so the docs REALLY ARE very outdated

TheTrigg committed:
I... Don't exactly know how I did that

TheTrigg committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into trigg-sucks-at-javascript

TheTrigg committed:
these will be manually set until I figure out how to make them scale like one without BREAKING FLEX

TheTrigg committed:
that was surprisingly easy

TheTrigg committed:
bruh

TheTrigg committed:
bit of cleanup

TheTrigg committed:
we don't use diagonal sprites

TheTrigg committed:
more cleanup; fixed renaming
Y'know that entire code block with the Carn comment?
Yeah, it wasn't actually being called for god knows how long.
IT WAS ALWAYS SUPPOSED TO LOG PEOPLE BEING RENAMED,
BUT IT NEVER DID

TheTrigg committed:
extra sanity checks

TheTrigg committed:
a tiny bit less fucky

TheTrigg committed:
the great de-trigg-ifying

TheTrigg committed:
of course I forgot a tiny fkn detail

TheTrigg committed:
rebuild tgui for the first time in 38 commits

TheTrigg committed:
sigh *resolves merge conflict*

TheTrigg committed:
bit more cleanup

TheTrigg committed:
more polish; organ damage buttons
...since that health manipulation panel won't happen in a while

TheTrigg committed:
this option does nothing
since it was ported in half by... me. whoops.

TheTrigg committed:
guess what, more polish

TheTrigg committed:
tiny stuff and rebuild

TheTrigg committed:
aaand flags because I forgot that they existed

TheTrigg committed:
one day I'll remember to rebuild tgui

TheTrigg commented:
So. This PR has been open for a while. I did say I'm slow.

Unless I forgot about something, feature parity with the old VV has indeed been reached.
Meaning... This can be testmerged?
But just in case, the verb to open the old VV is still there, also for admins to be able to compare the two panels. For now.
After the testmerge phase passes I plan on making it a hidden verb but keeping the code there

I will mark this as ready for review, even though not all on the todo list has been done.
I'm honestly burned out from smashing my head against JavaScript and this project as a whole. Hope you can understand why.

The code... Frontend and Backend... As said in the PR description, you'll probably want to gouge your eyes out.
As a relative starter to Tgui, DM, and coding in general, I'll excuse myself with "I don't know any better" because... Well, I don't.
I mean, yes, I could read other code and see how the big boys do it and stuff but... Again, I'm tired and just wanna move on from this PR.
I'll of course do any changes reviewers ask me to, if they get to that point instead of going "nah this code is too shit \*close pr*"
In which case, well, what can I do, really. ~~Maybe code decently you lazy sack of garbage~~

TheTrigg committed:
sorry for the notif, crossed

francinum commented on code/modules/tgui/states/admin.dm:
```diff
@@ -10,3 +10,17 @@ GLOBAL_DATUM_INIT(admin_state, /datum/ui_state/admin_state, new)
+GLOBAL_DATUM_INIT(VV_state, /datum/ui_state/VV_state, new)
+
+/datum/ui_state/VV_state/can_use_topic(src_object, mob/user)
+	if(check_rights_for(user.client, R_ADMIN))
```
This should be a rights check for `R_VAREDIT`. Not `R_ADMIN`.

In the current state coders are unable to use the panel to modify variables as they aren't given administrative rights.
This is also a violation of feature parity as it makes the UI completely uninteractable, including dropdown verbs.

TheTrigg replied:
yyyyeah my mind completely skipped thinking about permission stuff.

So I decided to turn this into just a check for holder, since in the old VV people with R_ADMIN could at least browse vars, but not edit and whatnot. Now permissions are checked respective to each action, except for where I missed adding a permission check, which is very probable to be at least one case.

TheTrigg committed:
permissions, cleanup, somehow a performance hit?

ike709 commented on code/modules/tgui/states/admin.dm:
```diff
@@ -14,13 +14,14 @@ GLOBAL_DATUM_INIT(admin_state, /datum/ui_state/admin_state, new)
 
 /datum/ui_state/VV_state/can_use_topic(src_object, mob/user)
-	if(check_rights_for(user.client, R_ADMIN))
+	if(check_rights_for(user.client))
```
???

TheTrigg replied:
Leaving the second argument empty makes the proc only check for holder but... I don't know why my dumb ass didn't just `if(user.client.holder)`

ike709 commented:
Untestmerged until permissions issues are sorted out.

TheTrigg committed:
I'm smart

francinum commented:
Horribly uncomfortable to use, has issues with nested lists going off the right edge, header images are oversized and can at times deform horrifyingly.

TheTrigg committed:
line endings be like "imma conflict"

TheTrigg committed:
upstream

TheTrigg committed:
thank you VSC for the invisible unsolved conflicts

ghost commented:
sooo, is this ever gonna be finished?

TheTrigg committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into trigg-sucks-at-javascript

TheTrigg committed:
byond why are you like this

TheTrigg committed:
this too

TheTrigg committed:
oh

TheTrigg commented:
Ohhh, right. This PR is still open.

So. After reading this I was immediately prompted to open up VSC and take a shot at fixing most issues pointed out to me. (like, a month ago)
I _kind of_ got rid of the lag issue after encountering some very weird byond behavior.
But then that broke a bunch of other things. Mostly lists.

Then I realized that I'll need to remake the backend instead of attempting to adapt it from the old VV.
But then I said "fuck no."
So yes I give up. Maybe someone will decide to continue this in the future. Maybe myself.

But now I'm closing this. I'm too much of a smoothbrain to make all this work correctly.
